### PR TITLE
Add new Complementarity formualtion for VLE with cubic EoSs

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
This PR adds an implementation of the improved Smooth VLE formulation originally proposed in #977. This implementation preserves the old implementation for backward compatibility.

## Changes proposed in this PR:
- Infrastructure work to support new formulation, and address some pylint issues in modular properties code.
- Add new class `SmoothVLE2` to implement improved formulation

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/models/properties/modular_properties/eos/ceos.py [hunk 1]
```diff
@@ -1356,26 +1382,79 @@ def a(k):
...
+        expression for fw
+
+    """
     return 0.48 + 1.574 * cobj.omega - 0.176 * cobj.omega**2
 
 
 def func_alpha_soave(T, fw, cobj):
+    """
+    alpha function for SRK EoS.
+
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/eos/ceos.py:1415]: I believe this alpha function is used by both SRK and PR.

File: idaes/models/properties/modular_properties/eos/ceos.py [hunk 2]
```diff
@@ -1356,26 +1382,79 @@ def a(k):
...
+    """
     Tc = cobj.temperature_crit
     Tr = T / Tc
     return (1 + fw * (1 - sqrt(Tr))) ** 2
 
 
 def func_dalpha_dT_soave(T, fw, cobj):
+    """
+    Function to get first derivative of alpha for SRK EoS.
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/eos/ceos.py:1431]: Same thing here.

File: idaes/models/properties/modular_properties/eos/ceos.py [hunk 3]
```diff
@@ -1412,8 +1497,14 @@ def rule_am_crit_default(m, cname, a_crit):
...
 def rule_bm_default(m, b, p):
+    """
+    Standard mixing rule for b term
+    """
     return sum(m.mole_frac_phase_comp[p, i] * b[i] for i in m.components_in_phase(p))
 
 
 def rule_bm_crit_default(m, b):
+    """
+    Standard mixing rule for b term at critical point
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/eos/ceos.py:1508]: "at critical point" -> "evaluated at critical point"

File: idaes/models/properties/modular_properties/examples/tests/test_BT_PR_SmoothVLE2.py [hunk 4]
```diff
@@ -0,0 +1,891 @@
...
+        m.fs.state[1].temperature.setub(600)
+
+        for logP in [9.5, 10, 10.5, 11, 11.5, 12]:
+            m.fs.obj.deactivate()
+
+            m.fs.state[1].flow_mol.fix(100)
+            m.fs.state[1].mole_frac_comp["benzene"].fix(0.5)
+            m.fs.state[1].mole_frac_comp["toluene"].fix(0.5)
+            m.fs.state[1].temperature.fix(300)
+            m.fs.state[1].pressure.fix(10 ** (0.5 * logP))
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/examples/tests/test_BT_PR_SmoothVLE2.py:219]: What's the point of the 0.5 in the exponent here?

File: idaes/models/properties/modular_properties/examples/tests/test_BT_PR_SmoothVLE2.py [hunk 5]
```diff
@@ -0,0 +1,891 @@
...
+        assert (
+            m.fs.state[1].scaling_factor[m.fs.state[1].mole_frac_comp["toluene"]]
+            == 1000
+        )
+        assert (
+            m.fs.state[1].scaling_factor[
+                m.fs.state[1].mole_frac_phase_comp["Liq", "benzene"]
+            ]
+            == 1000
+        )
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/examples/tests/test_BT_PR_SmoothVLE2.py:836]: For the record, I don't think 1000 is a great value for scaling factors of mole fractions. Testing mole fractions like this makes it a pain to change later. Can't we do something like test  Jacobian rows, columns, and condition number?

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+            henry_comps,
+            l_only_comps,
+            v_only_comps,
+        ) = identify_VL_component_list(b, phase_pair)
+
+        if l_phase is None or v_phase is None:
+            raise ConfigurationError(
+                f"{b.params.name} Generic Property Package phase pair {phase_pair[0]}-{phase_pair[1]} "
+                "was set to use Smooth VLE formulation, however this is not a vapor-liquid pair."
+            )
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:61]: This error should probably be thrown by `identify_VL_component_list`.

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+                liquid_phase = phase_pair[1]
+            else:
+                vapor_phase = phase_pair[1]
+                liquid_phase = phase_pair[0]
+            return (
+                b._teq[phase_pair]
+                - b.temperature
+                - s[vapor_phase] * t_units
+                + s[liquid_phase] * t_units
+                == 0
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:90]: Why make the slack variable `s` dimensionless when we're multiplying it by `t_units` here? That also implies it should be scaled.

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+        b.add_component("_tbar_constraint" + suffix, Constraint(rule=rule_teq))
+
+        eps = Param(
+            default=1e-04,
+            mutable=True,
+            doc="Smoothing parameter for complementarities",
+            units=f_units,
+        )
+        b.add_component("eps" + suffix, eps)
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:101]: Does `eps` need to be adjusted by the user?

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+            bounds=(0, None),
+            doc="Slack variable for cubic second derivative for phase p",
+            units=pyunits.dimensionless,
+        )
+        b.add_component("gn" + suffix, gn)
+
+        def rule_temperature_slack_complementarity(b, p):
+            flow_phase = b.flow_mol_phase[p]
+
+            return smooth_min(s[p] * f_units, flow_phase, eps) == 0
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:124]: Okay, we're treating `s` as having different units depending on the context. That's not ideal for scaling.

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        T_units = blk.params.get_metadata().default_units.TEMPERATURE
+
+        liquid_phase, _, raoult_comps, henry_comps, _, _ = identify_VL_component_list(
+            blk, pp
+        )
+
+        Tbub = estimate_Tbub(blk, T_units, raoult_comps, henry_comps, liquid_phase)
+        Tdew = estimate_Tbub(blk, T_units, raoult_comps, henry_comps, liquid_phase)
+
+        assert False
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:187]: Debugging survival into production.

**Review by dallan-keylogic — COMMENTED** [review]
Here are some comments, it seems like the full PR isn't finished yet.

File: idaes/models/properties/modular_properties/base/utility.py [hunk 11]
```diff
@@ -288,3 +301,293 @@ def get_concentration_term(blk, r_idx, log=False):
...
+                .dT_expression(blk, liquid_phase, j, Tbub0 * T_units)
+                for j in henry_comps
+            )
+        )
+
+        # Limit temperature step to avoid excessive overshoot
+        if f / df < -50:
+            Tbub1 = Tbub0 + 50
+        elif f / df > 50:
+            Tbub1 = Tbub0 - 50
```
[the remark this answers is not in the document]
Comment (dallan-keylogic) [reply idaes/models/properties/modular_properties/base/utility.py:444]: Probably too much work, though.

File: idaes/models/properties/modular_properties/base/utility.py [hunk 12]
```diff
@@ -288,3 +301,293 @@ def get_concentration_term(blk, r_idx, log=False):
...
+    Returns:
+        Lists of component names for:
+            * liquid Phase object
+            * vapor Phase object
+            * components using Raoult's Law
+            * components using Henry's Law
+            * liquid only components,
+            * vapor only components
+
+    """
```
[the remark this answers is not in the document]
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/base/utility.py:323]: This is only for VLE systems, so is not expected to work in those cases. I have added an error for cases where a non-VL pair is passed as the argument.

File: idaes/models/properties/modular_properties/base/utility.py [hunk 13]
```diff
@@ -288,3 +301,293 @@ def get_concentration_term(blk, r_idx, log=False):
...
+    # Use lowest component temperature_crit as starting point
+    # Starting high and moving down generally works better,
+    # as it under-predicts next step due to exponential form of
+    # Psat.
+    # Subtract 1 to avoid potential singularities at Tcrit
+    Tbub0 = (
+        min(blk.params.get_component(j).temperature_crit.value for j in raoult_comps)
+        - 1
```
[the remark this answers is not in the document]
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/base/utility.py:405]: In this case, `1` is just an arbitrary value to move away from the exact critical value as some correlations are singular if `T=Tc[j]`.

File: idaes/models/properties/modular_properties/base/utility.py [hunk 14]
```diff
@@ -288,3 +301,293 @@ def get_concentration_term(blk, r_idx, log=False):
...
+        Estimated dew point pressure as a float.
+
+    """
+    return value(
+        1
+        / (
+            sum(blk.mole_frac_comp[j] / blk.pressure_sat_comp[j] for j in raoult_comps)
+            + sum(
+                blk.mole_frac_comp[j] / blk.henry[liquid_phase, j] for j in henry_comps
+            )
```
[the remark this answers is not in the document]
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/base/utility.py:600]: Maybe - but that would be a definite case of user stupidity.

File: idaes/models/properties/modular_properties/examples/tests/test_BT_PR_SmoothVLE2.py (hunk 4 above)
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/examples/tests/test_BT_PR_SmoothVLE2.py]: To get a half step size, as `range` does not accept non-integer arguments.

File: idaes/models/properties/modular_properties/examples/tests/test_BT_PR_SmoothVLE2.py (hunk 5 above)
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/examples/tests/test_BT_PR_SmoothVLE2.py]: Agreed, but for now I am focused on getting this working (plus the new tools will supersede this before too long).

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py (hunk 6 above)
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:74]: A catch has been added to the utility method, but this can be retained as well.

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py (hunk 8 above)
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:115]: Yes, users should adjust `eps` as necessary and this will be in the docs (the old formulation was the same).

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py (hunk 9 above)
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:138]: Agreed - if you have any ideas on how to handle this better it would be appreciated. For now I have just focused on getting the code running.

**Comment by andrewlee94** [comment]
@viiibhav @jghouse88 Not sure if either of you still receive these, but input and review on this PR would be welcome.

File: idaes/models/properties/modular_properties/base/utility.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dallan-keylogic) [reply idaes/models/properties/modular_properties/base/utility.py:405]: Yeah, but in edge cases we could end up with a negative temperature (for example, if the user decided to work with temperature reduced by `Tc`).

File: idaes/models/properties/modular_properties/examples/tests/test_BT_PR_SmoothVLE2.py (hunk 4 above)
Comment (dallan-keylogic) [reply idaes/models/properties/modular_properties/examples/tests/test_BT_PR_SmoothVLE2.py]: Then it's not really `logP` anymore, but `2*logP`, which isn't a great variable name. You could get around this issue with `numpy.linspace` or `numpy.logspace`, but that's probably not worth the effort changing.

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py (hunk 9 above)
Comment (dallan-keylogic) [reply idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:138]: My instinct here is to just give `s[p]` units of temperature per flowrate, then scale this equation like temperature. However, it needs to be scaled with a "temperature difference" scaling factor, not an "absolute temperature" scaling factor.

**Comment by jghouse88** [comment]
> @viiibhav @jghouse88 Not sure if either of you still receive these, but input and review on this PR would be welcome.

Will take a look this week if that works. The best test would be to see if the paper examples work with this pr and see if we get similar results.

**Comment by andrewlee94** [comment]
@jghouse88 I have tested this with the existing tests in IDAES, swapping the new formulation for the old one. All but two tests pass with the new formulation, but the two that fail appear to need further investigation - the temperature sweep test is failing to converge to the correct solution (it gets stuck somewhere).

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py (hunk 9 above)
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py]: Based on some testing, scaling of s, g and epsilon depends on the larger of T and flow, thus it is hard to know what to do with these. I am inclined to leave these as unit-less for that reason, and any scaling routine can handle the necessary logic behind the scenes (as I doubt users will care enough to understand what is happening).

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+        vl_phase_set = Set(initialize=[phase_pair[0], phase_pair[1]])
+        b.add_component("_vle_set" + suffix, vl_phase_set)
+
+        s = Var(
+            vl_phase_set,
+            initialize=EPS_INIT,
+            bounds=(0, None),
+            doc="Slack variable for equilibrium temperature",
+            units=t_units,
+        )
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:115]: Do we have any sort of scaling for these slack variables, or is the intention to let the autoscaler do it? This is a case where the results from the autoscaler is going to be highly dependent on initialization, and since the slack variable will frequently be near zero, you may not get any scaling at all.

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+        Method to calculate scaling factors for phase equilibrium
+        """
+        suffix = "_" + phase_pair[0] + "_" + phase_pair[1]
+        sf_T = iscale.get_scaling_factor(b.temperature, default=1, warning=True)
+
+        try:
+            teq_cons = getattr(b, "_teq_constraint" + suffix)
+            # pylint: disable-next=protected-access
+            iscale.set_scaling_factor(b._teq[phase_pair], sf_T)
+            iscale.constraint_scaling_transform(teq_cons, sf_T, overwrite=False)
```
Comment (dallan-keylogic) [inline idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:227]: I guess we're not scaling the slacks then. If we're giving them temperature units, we might as well scale them like temperature. It would be a good first guess, at any rate, and I think the autoscalers will faceplant.

**Review by dallan-keylogic — CHANGES_REQUESTED** [review]

File: docs/explanations/components/property_package/general/pe/smooth_vle2.rst [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+.. math:: \min(a, b) =  0.5{\left[a + b - \sqrt{(a-b)^2 + \epsilon^2}\right]}
+
+Each complementarity requires a smoothing parameter, named :math:`\epsilon_T` and :math:`\epsilon_Z` for the temperature and cubic root constraints respectively. Within the IDAES model, these are rendered as ``eps_t_phase1_phase2`` and ``eps_z_phase1_phase2``, where ``phase1`` and ``phase2`` are the names assigned to the liquid and vapor phases in the property package (order will depend on the order these are declared).
+
+The tractability of the VLE problem depends heavily upon the values chosen for :math:`\epsilon_T` and :math:`\epsilon_Z`, with larger values resulting in smoother transitions at the phase boundaries (and thus increased tractability) at the expense of decreased accuracy near these points. It is recommended that users employ a 2-stage approach to solving these problems, starting with a larger value of :math:`\epsilon_T` and :math:`\epsilon_Z` initially to determine which region the solution lies in, followed by a second solve using smaller values to refine the solution.
```
[the remark this answers is not in the document]
Comment (andrewlee94) [reply docs/explanations/components/property_package/general/pe/smooth_vle2.rst:64]: They are local, as they potentially need to vary unit-to-unit. Their names are show in the docs.

File: idaes/models/properties/modular_properties/base/generic_property.py [hunk 18]
```diff
@@ -2424,9 +2097,13 @@ def build(self):
...
 
             t_units = self.params.get_metadata().default_units.TEMPERATURE
+            if self.temperature.value is not None:
+                t_value = value(self.temperature)
+            else:
+                t_value = None
             self._teq = Var(
                 self.params._pe_pairs,
-                initialize=value(self.temperature),
+                initialize=t_value,
```
[the remark this answers is not in the document]
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/base/generic_property.py:2106]: This is a catch for cases where T might not be a state variable. I do not expect it to happen often, but it can occur so we need to catch it.

File: idaes/models/properties/modular_properties/examples/tests/test_BT_PR.py [hunk 19]
```diff
@@ -80,23 +85,34 @@ def test_T_sweep(self, m):
...
             m.fs.state[1].mole_frac_comp["benzene"].fix(0.5)
             m.fs.state[1].mole_frac_comp["toluene"].fix(0.5)
             m.fs.state[1].temperature.fix(300)
-            m.fs.state[1].pressure.fix(10 ** (0.5 * logP))
+            m.fs.state[1].pressure.fix(P)
+
+            # For optimization sweep, use a large eps to avoid getting stuck at
+            # bubble and dew points
+            m.fs.state[1].eps_t_Vap_Liq.set_value(10)
+            m.fs.state[1].eps_z_Vap_Liq.set_value(10)
```
[the remark this answers is not in the document]
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/examples/tests/test_BT_PR.py:99]: Whilst there is potentially more care needed with the eps values, it is not much more (and the fact that we did not need to tweak the eps values for the older implementation is likely just as much luck as anything). However, I would argue "Robust" is no the correct word in this case; both methods are sensitive to the value of eps chosen, but once a good value has been identified they will generally be robust to changes.

File: idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py (hunk 16 above)
Comment (dallan-keylogic) [reply idaes/models/properties/modular_properties/phase_equil/smooth_VLE_2.py:227]: Very well, but I do expect this case will need heuristic scaling.

**Review by dallan-keylogic — APPROVED** [review]
LGTM

File: idaes/models/properties/modular_properties/base/generic_property.py [hunk 20]
```diff
@@ -2424,9 +2097,13 @@ def build(self):
         ):
 
             t_units = self.params.get_metadata().default_units.TEMPERATURE
+            if self.temperature.value is not None:
+                t_value = value(self.temperature)
+            else:
+                t_value = None
```
Comment (bpaul4) [inline idaes/models/properties/modular_properties/base/generic_property.py:2103]: If this is None, does the model break or is there a catch for this case?

File: idaes/models/properties/modular_properties/base/tests/test_generic_property.py [hunk 21]
```diff
@@ -1972,6 +1975,57 @@ def build_critical_properties(b, *args, **kwargs):
...
+    "state_bounds": {
+        "flow_mol": (0, 100, 1000, pyunits.mol / pyunits.s),
+        "temperature": (273.15, 300, 500, pyunits.K),
+        "pressure": (5e4, 1e5, 1e6, pyunits.Pa),
+    },
+    "pressure_ref": (101325, pyunits.Pa),
+    "temperature_ref": (298.15, pyunits.K),
+    "parameter_data": {
+        "PR_kappa": {
+            ("foo", "bar"): 0.000,
```
Comment (bpaul4) [inline idaes/models/properties/modular_properties/base/tests/test_generic_property.py:2023]: Why "foo" and "bar"?

File: idaes/models/properties/modular_properties/base/tests/test_utility.py [hunk 22]
```diff
@@ -559,3 +581,253 @@ def test_inherent_partial_pressure(self, frame2):
...
+    class DummyParameterData(GenericParameterData):
+        def configure(self):
+            self.configured = True
+
+    @pytest.fixture()
+    def frame(self):
+        m = ConcreteModel()
+        m.params = DummyParameterBlock(
+            components={
+                "a": {},
```
Comment (bpaul4) [inline idaes/models/properties/modular_properties/base/tests/test_utility.py:613]: Are these supposed to be empty?

**Review by bpaul4 — APPROVED** [review]
Looks good, just a couple minor questions.

File: idaes/models/properties/modular_properties/base/generic_property.py (hunk 20 above)
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/base/generic_property.py:2103]: This just affects whether we can get initial guesses for Teq. Whether the model will solve or not is a separate issue at this point; the goal here is for the code to work without encountering a terminal error.

File: idaes/models/properties/modular_properties/base/tests/test_generic_property.py (hunk 21 above)
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/base/tests/test_generic_property.py:2023]: [Standard Python terms for testing attributes.](https://www.altcademy.com/explain/foo-bar-baz)

File: idaes/models/properties/modular_properties/base/tests/test_utility.py (hunk 22 above)
Comment (andrewlee94) [reply idaes/models/properties/modular_properties/base/tests/test_utility.py:613]: Yes - this is a unit test. We need to define the components so the code runs but don't care about any actual properties.
